Update uses of `PAYMENT_STORE_KEY` to use the `StoreDescriptor`

opr opened the pull request:
### Changes proposed in this Pull Request:

This PR updates uses of `PAYMENT_STORE_KEY` to use the `paymentStore` store descriptor instead.

It also updates an instance of `core/notices` being used and uses the store descriptor of `core/notices` (`noticesStore`) instead.

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #54427

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

The main thing to test is that several different payment methods work correctly with the Cart/Checkout blocks. You should test both regular and express payment methods and ensure there are no JS errors during the checkout process.

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### `usePaymentMethodState` and `usePaymentMethodInterface`

1. Install and activate Stripe. Activate some other payment methods.
2. Go to the Checkout block, ensure you can switch between payment methods.
3. Check out using [one of Stripe's failed cards](https://docs.stripe.com/testing#declined-payments) .
4. Ensure the errors prevent you from checking out.
5. Ensure you can check out with a valid card.

#### `PaymentEventsProvider`

1. Edit `plugins/woocommerce-blocks/assets/js/extensions/payment-methods/bacs/index.js` to the following:

```js
/**
 * External dependencies
 */
import { registerPaymentMethod } from '@woocommerce/blocks-registry';
import { __ } from '@wordpress/i18n';
import { getPaymentMethodData } from '@woocommerce/settings';
import { decodeEntities } from '@wordpress/html-entities';
import { sanitizeHTML } from '@woocommerce/utils';
import { RawHTML } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { PAYMENT_METHOD_NAME } from './constants';

const settings = getPaymentMethodData( 'bacs', {} );
const defaultLabel = __( 'Direct bank transfer', 'woocommerce' );
const label = decodeEntities( settings?.title || '' ) || defaultLabel;

/**
 * Content component
 */
const Content = ( { paymentStatus } ) => {
	console.log( paymentStatus );
	return <RawHTML>{ sanitizeHTML( settings.description || '' ) }</RawHTML>;
};

/**
 * Label component
 *
 * @param {*} props Props from payment API.
 */
const Label = ( props ) => {
	const { PaymentMethodLabel } = props.components;
	return <PaymentMethodLabel text={ label } />;
};

/**
 * Bank transfer (BACS) payment method config object.
 */
const bankTransferPaymentMethod = {
	name: PAYMENT_METHOD_NAME,
	label: <Label />,
	content: <Content />,
	edit: <Content />,
	canMakePayment: () => true,
	ariaLabel: label,
	supports: {
		features: settings?.supports ?? [],
	},
};

registerPaymentMethod( bankTransferPaymentMethod );
```

7. In WooCommerce -> Settings -> Payments, enable BACS.
8. Install and enable Stripe.
9. Open the JS Console.
10. Add an item to your cart and go to the Checkout block.
11. Check "preserve log" in the JS console.
12. Complete checkout, and ensure you see the status (logged in `bacs/index.js`) change. Don't worry about it changing to all statuses, just ensure it logs without errors and changes.

#### `useCheckoutSubmit`

1. Ensure you can check out using an express payment method. While the express method is open, ensure the "Place order" button at the bottom of the checkout form is disabled.
2. Pay using an express payment method on Cart block, ensure the express methods appear disabled while you're interacting with one.

#### `CheckoutEventsProvider` and `updatePaymentMethods`

1. Go to WooCommerce -> Settings -> Payments and enable BACS.
3. Edit `plugins/woocommerce-blocks/assets/js/extensions/payment-methods/bacs/index.js`
4. Change the `canMakePayment` method to

```js
( { cart } ) => {
		return cart.billingAddress.country === 'US';
	}
```

5. Load the Checkout block and change the billing country to US. 
6. Ensure you see the BACS method as an available payment method.
7. Change it to a different country, ensure it disappears.
8. Change it back to US and ensure you can check out using BACS.

#### `PaymentMethodCard`

1. Install and activate Stripe.
2. As a logged in user, add an item to your cart and go to the Checkout block.
3. Pay using Stripe, and ensure you can save the method to your account. 
4. Check the box to save the method to your account, uncheck it and verify the state can be toggled.
5. Check it again, and check out.
6. Re-add an item to your cart and go back to the Checkout block.
7. Verify the method you just paid with is saved and can be used to check out again.

#### Incompatible payment method notice

1. Install an incompatible payment method (e.g. https://wordpress.org/plugins/express-checkout/)
2. Go to the Checkout block in the editor.
3. Select the Checkout Payments block.
4. Ensure you see a warning in the sidebar about the incompatible payment method.

#### Suppressing notices while express payment is active

1. Install and activate Stripe and set up the express payment methods.
2. Add a notice to the site by pasting this code into your JS console: `wp.data.dispatch( wp.notices.store ).createErrorNotice( 'Test notice', { context: 'wc/checkout' } );`
3. Start paying using the express payment methods.
4. Ensure the notice disappears.
5. Exit the express payment method without paying.
6. Ensure the notice returns.


<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

opr committed:
Update to use store descriptor in payment selectors test

opr committed:
Update to use store descriptor in checkout flow docs

opr committed:
Update to use store descriptor in PAymentEventsProvider

opr committed:
Update to use store descriptor in usePaymentMethodInterface

opr committed:
Update to use store descriptor in useCheckoutSubmit

opr committed:
Update to use store descriptor in useCheckoutSubmit tests

opr committed:
Update to use store descriptor in usePaymentMethods

opr committed:
Update docs

opr committed:
Update to use store descriptor in setDefaultPaymentMethods

opr committed:
Update to use store descriptor in CheckoutEventsProvider

opr committed:
Update to use store descriptor in updatePaymentMethods

opr committed:
Update to use store descriptor in CheckoutProcessor

opr committed:
Export store config in separate file

opr committed:
Update to use store descriptor in express payment methods

opr committed:
Update comment in payment method registry

opr committed:
Update to use store descriptor in PaymentMethodCard

opr committed:
Update to use store descriptor in CheckoutExpressPayment

opr committed:
Update to use store descriptor in PaymentMethodOptions

opr committed:
Update to use store descriptor in PaymentMethods

opr committed:
Update to use store descriptor in SavedPaymentMethodOptions

opr committed:
Update to use store descriptor in ExpressPaymentMethods

opr committed:
Update to use store descriptor in payment method tests

opr committed:
Update to use store descriptor in checkout payment block

opr committed:
Update to use store descriptor in tests

opr committed:
Update to use store descriptor in StoreNoticesContainer

opr committed:
Update to use store descriptor in checkPaymentMethods

opr committed:
Update to use store descriptor in useIncompatiblePaymentGatewaysNotice

opr committed:
Update to use store descriptor in Checkout thunks

opr committed:
Remove store-config file and revert back to config in index

opr committed:
Update misconfigured tests

opr committed:
Export payment store config type

opr committed:
Import payment store types into registry

This is because importing the store descriptor directly results in a circular dependency

opr committed:
Add changelog

opr commented on plugins/woocommerce-blocks/assets/js/blocks-registry/payment-methods/registry.ts:
```diff
@@ -11,15 +11,20 @@ import type {
 import { dispatch } from '@wordpress/data';
+import {
+	ActionCreatorsOf,
+	ConfigOf,
+	DispatchReturn,
+} from '@wordpress/data/build-types/types';
 
 /**
  * Internal dependencies
  */
 import { default as PaymentMethodConfig } from './payment-method-config';
 import { default as ExpressPaymentMethodConfig } from './express-payment-method-config';
 import { canMakePaymentExtensionsCallbacks } from './extensions-config';
-
-import { STORE_KEY as PAYMENT_STORE_KEY } from '../../data/payment/constants'; // Full path here because otherwise there's a circular dependency.
+import { STORE_KEY as PAYMENT_STORE_KEY } from '../../data/payment/constants';
+import type { PaymentStoreDescriptor } from '../../data/payment';
```
We can't get `paymentStore` from `@woocommerce/block-data` here as it introduces a circular dependency. The registry is really tightly coupled with the payment data store, so untangling these to remove the circular dependency would be too much work. Instead, I am importing only the _type_ of the `paymentStore` and casting the type returned by `dispatch`. It's the same result, just a little messier. This is the only file I had to do this in.

opr committed:
Update createNotice to use store descriptors

senadir reviewed (changes requested):
I didn't go too deep into testing payments (but I did test core ones), there's a console error in Checkout on load:

<img width="464" alt="image" src="https://github.com/user-attachments/assets/521f05a5-38dd-46bd-b603-a21cbbbe1f61" />

opr committed:
Add docs manifest

opr committed:
Revert "Update createNotice to use store descriptors"

This reverts commit 41166e7d09735cd27e3c9c17737a05d223de0a5c.

opr committed:
Export store notices store descriptor

opr committed:
Type cast selects to their CurriedSelectorsOf type

senadir commented:
it seems this change caused a 35kb increase (gzipped, so like 120kb) in the bundle size, is this expected?

opr commented:
> it seems this change caused a 35kb increase (gzipped, so like 120kb) in the bundle size, is this expected?

Unfortunately, yes, it's because we are now importing the store descriptor (so, actions, selectors etc.) rather than just the data store key string.

senadir commented:
Shouldnt those be already on another global script instead of being duplicated? is it worth it to add this load to skip a TS warning?

opr committed:
Prevent importing entire data store in express-methods

opr committed:
Fix unused dependency

opr commented:
> 9b5f562

Yeah, in [9b5f562](https://github.com/woocommerce/woocommerce/pull/54685/commits/9b5f562ffa0da572f2b9ee33c5f8758bc021b7bf) I removed the dependency from express payment methods, and reverted to using the store key with type casting. I think this is fine, it's something we can't get around because of a circular dependency.

senadir commented on plugins/woocommerce-blocks/assets/js/base/utils/create-notice.ts:
```diff
@@ -33,14 +39,16 @@ export const createNotice = (
 	const noticeContext = options?.context;
-	const suppressNotices =
-		select( 'wc/store/payment' ).isExpressPaymentMethodActive();
+	const selectors = select(
+		'wc/store/payment'
+	) as CurriedSelectorsOf< PaymentStoreDescriptor >;
+	const suppressNotices = selectors.isExpressPaymentMethodActive();
```
So this is the fix? nice.

opr replied:
Yeah, not ideal, it's much nicer to get it from the store descriptor, but given the circular dependencies we can't, and this is what we'd get back anyway.

We should avoid doing this _where possible_ because it's harder to manage if the types change in future.

senadir reviewed (approved):
Testing and looking good thank you Thomas
